frontend/z3_axioms.py

```diff
@@ -0,0 +1,214 @@
+    lst = [value == List(target)]
+    if position == 1:
+        t = [
+            value == Tuple1(target),
+            Exists(x, value == Tuple2(target, x), patterns=[Tuple2(target, x)]),
+            Exists([x, y], value == Tuple3(target, x, y), patterns=[Tuple3(target, x, y)]),
+            Exists([x, y, z], value == Tuple4(target, x, y, z), patterns=[Tuple4(target, x, y, z)]),
+            Exists([x, y, z, l], value == Tuple5(target, x, y, z, l), patterns=[Tuple5(target, x, y, z, l)]),
+        ]
+    elif position == 2:
+        t = [
+            Exists(x, value == Tuple2(x, target)),
+            Exists([x, y], value == Tuple3(x, target, y)),
+            Exists([x, y, z], value == Tuple4(x, target, y, z)),
+            Exists([x, y, z, l], value == Tuple5(x, target, y, z, l)),
+        ]
+    elif position == 3:
+        t = [
+            Exists([x, y], value == Tuple3(x, y, target)),
+            Exists([x, y, z], value == Tuple4(x, y, target, z)),
+            Exists([x, y, z, l], value == Tuple5(x, y, target, z, l)),
+        ]
+    elif position == 4:
+        t = [
+            Exists([x, y, z], value == Tuple4(x, y, z, target)),
+            Exists([x, y, z, l], value == Tuple5(x, y, z, target, l)),
+        ]
+    elif position == 5:
+        t = [
+            Exists([x, y, z, l], value == Tuple5(x, y, z, l, target)),
+        ]
+    else:
+        raise NotImplementedError("Inference for tuple assignment with length > 5 is not supported.")
+
+    return [Or(lst + t)]
+
+
+def index_assignment(indexed, index, value):
+    return [
+        Or(
+            indexed == Dict(index, value),
+            And(subtype(index, Int), indexed == List(value))
+        )
+    ]
+
+
+def slice_assignment(lower, upper, step, sliced, value):
+    return [
+        And(subtype(lower, Int), subtype(upper, Int), subtype(step, Int), subtype(sliced, Seq),
+            Exists([x], And(sliced == List(x), value == List(x))))
+    ]
+
+
+def delete_subscript(indexed):
+    return [
+        Not(Or(
+            indexed == String,
+            indexed == Bytes,
+            subtype(indexed, Tuple)
+        ))
+    ]
+
+
+def body(result, new):
+    return [
+        Implies(new != zNone, subtype(new, result))
+    ]
+
+
+def control_flow(body, orelse, result):
+    # TODO numeric casting
+    return [
+        Implies(orelse == zNone, result == body),
+        Implies(orelse != zNone, And(
+            subtype(body, result),
+            subtype(orelse, result)
+        ))
+    ]
+
+
+def for_loop(iterable, target):
+    return [
+        Or(
+            iterable == List(target),
+            iterable == Set(target),
+            Exists([x], iterable == Dict(target, x), patterns=[Dict(target, x)]),
+            And(iterable == String, target == String),
+            And(iterable == Bytes, target == Bytes)
+        )
+    ]
+
+
+def try_except(body, orelse, final, result):
+    return [
+        subtype(body, result),
```

Comment:
If only some parts return something then the other parts will have a type None, and the solver will choose the common super-type between None and the inferred type (which in most cases will be object type).